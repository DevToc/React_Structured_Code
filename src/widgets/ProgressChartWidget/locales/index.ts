import EN_PROGRESS_CHART_TOOLBAR from './en/toolbar.json';
import ES_PROGRESS_CHART_TOOLBAR from './es/toolbar.json';
import FR_PROGRESS_CHART_TOOLBAR from './fr/toolbar.json';

export { EN_PROGRESS_CHART_TOOLBAR, ES_PROGRESS_CHART_TOOLBAR, FR_PROGRESS_CHART_TOOLBAR };
