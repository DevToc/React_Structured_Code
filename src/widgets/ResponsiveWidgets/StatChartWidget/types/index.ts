import { StatChartType } from '../StatChartWidget.types';
import { DonutStatChart } from './DonutStatChart';
import { HalfDonutStatChart } from './HalfDonutStatChart';
import { ProgressBarStatChart } from './ProgressBarStatChart';
import { IconStatChart } from './IconStatChart';

export const CHART_TYPE_MAP = {
  [StatChartType.Donut]: DonutStatChart,
  [StatChartType.HalfDonut]: HalfDonutStatChart,
  [StatChartType.ProgressBar]: ProgressBarStatChart,
  [StatChartType.Icon]: IconStatChart,
};
