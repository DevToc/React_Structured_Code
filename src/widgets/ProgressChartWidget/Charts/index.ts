import { ProgressChartType } from '../ProgressChartWidget.types';
import { DonutChart } from './Donut';
import { BarChart } from './Bar';

export const CHART_TYPE_MAP = {
  [ProgressChartType.Donut]: {
    component: DonutChart,
    props: { type: 'full' },
  },
  [ProgressChartType.HalfDonut]: {
    component: DonutChart,
    props: { type: 'half' },
  },
  [ProgressChartType.Bar]: {
    component: BarChart,
    props: null,
  },
};
