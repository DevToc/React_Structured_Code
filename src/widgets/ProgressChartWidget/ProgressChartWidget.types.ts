import { Widget } from 'types/widget.types';

export enum ProgressChartType {
  Donut = 'Donut',
  HalfDonut = 'HalfDonut',
  Bar = 'Bar',
}

interface ProgressChartWidgetData extends Widget {
  type: ProgressChartType;

  // Percent in decimal form (0-1)
  // Note: this is the value that is directly passed to the Nivo Pie
  // component but the value is normalized when displayed in the Donut Size
  // toolbar slider due to issues rendering using the entire range of 0-100%
  donutSize: number;

  // Normalized value in range [5-50]
  // This is the value displayed in the toolbar
  barHeight: number;

  // 0-100
  value: number;

  // Percent in decimal form (0-1)
  cornerRadius: number;

  dataColor: string;
  nonDataColor: string;
}

export type { ProgressChartWidgetData };
