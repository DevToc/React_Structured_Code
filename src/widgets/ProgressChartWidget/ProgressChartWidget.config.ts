import { ProgressChartType } from './ProgressChartWidget.types';

const DEFAULT_VALUE = 67;
const DEFAULT_DATA_COLOR = '#0069D1';
const DEFAULT_NON_DATA_COLOR = '#D9E1FF';

export const DEFAULT_WIDTH = 150;
export const DEFAULT_HEIGHT = 75;
export const DEFAULT_BAR_HEIGHT = 12;

// Max and min value for bar height in toolbar
export const MAX_BAR_HEIGHT = 50;
export const MIN_BAR_HEIGHT = 5;

export const DEFAULT_PROGRESS_CHART_DATA_MAP = {
  [ProgressChartType.Donut]: {
    type: ProgressChartType.Donut,

    donutSize: 0.87,
    barHeight: DEFAULT_BAR_HEIGHT,
    value: DEFAULT_VALUE,
    cornerRadius: 1,
    dataColor: DEFAULT_DATA_COLOR,
    nonDataColor: DEFAULT_NON_DATA_COLOR,

    widthPx: DEFAULT_WIDTH,
    heightPx: DEFAULT_WIDTH,
  },
  [ProgressChartType.HalfDonut]: {
    type: ProgressChartType.HalfDonut,

    donutSize: 0.85,
    barHeight: DEFAULT_BAR_HEIGHT,
    value: DEFAULT_VALUE,
    cornerRadius: 1,
    dataColor: DEFAULT_DATA_COLOR,
    nonDataColor: DEFAULT_NON_DATA_COLOR,

    widthPx: DEFAULT_WIDTH,
    heightPx: DEFAULT_HEIGHT,
  },
  [ProgressChartType.Bar]: {
    type: ProgressChartType.Bar,

    donutSize: 0.85,
    barHeight: DEFAULT_BAR_HEIGHT,
    value: DEFAULT_VALUE,
    cornerRadius: 1,
    dataColor: DEFAULT_DATA_COLOR,
    nonDataColor: DEFAULT_NON_DATA_COLOR,

    widthPx: DEFAULT_WIDTH,
    heightPx: DEFAULT_BAR_HEIGHT,
  },
};
