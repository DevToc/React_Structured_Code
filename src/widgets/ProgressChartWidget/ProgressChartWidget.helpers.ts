import { WidgetType } from 'types/widget.types';
import { normalizeNumber } from 'utils/number';
import { DEFAULT_PROGRESS_CHART_DATA_MAP } from './ProgressChartWidget.config';
import { ProgressChartType, ProgressChartWidgetData } from './ProgressChartWidget.types';
import { VERSION } from './ProgressChartWidget.upgrade';

/**
 * Converts percentage input from toolbar Donut Size slider to value to be saved
 * in widget data. Normalizes the input.
 *
 * @param donutSize number This is an integer between [0, 100] that is displayed in the toolbar slider
 * @param type ProgressChartType
 * @returns decimal number to be saved in ProgressChart widget data
 */
export const convertDonutSizeInputToChartData = (donutSize: number, type: ProgressChartType) => {
  const normalized =
    type === ProgressChartType.HalfDonut
      ? normalizeNumber(60, 98, 0.38 * donutSize + 60)
      : normalizeNumber(0, 98, 0.98 * donutSize);

  return normalized / 100;
};

/**
 * Converts donutSize property of widget data to value that will be displayed in the
 * toolbar Donut Size slider input.
 *
 * @param donutSize number Donut size as a percent in decimal form - this is what is stored in the widget data
 * @param type ProgressChartType
 * @returns
 */
export const convertDonutSizeChartDataToInput = (donutSize: number, type: ProgressChartType) => {
  return type === ProgressChartType.HalfDonut ? (donutSize * 100 - 60) / 0.38 : (donutSize * 100) / 0.98;
};

export const generateDefaultData = (
  chartType: ProgressChartType,
  progressChartConfig?: Partial<ProgressChartWidgetData>,
) => {
  return Object.assign({
    widgetType: WidgetType.ProgressChart,
    widgetData: {
      version: VERSION,
      topPx: 0,
      leftPx: 0,
      rotateDeg: 0,
      isHidden: progressChartConfig?.isHidden || false,

      ...DEFAULT_PROGRESS_CHART_DATA_MAP[chartType],
    },
  });
};
