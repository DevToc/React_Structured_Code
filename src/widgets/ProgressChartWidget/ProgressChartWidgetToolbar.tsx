import { useMemo } from 'react';
import { Flex } from '@chakra-ui/react';

import { ProgressChartType, ProgressChartWidgetData } from './ProgressChartWidget.types';
import { ColorPicker } from 'modules/Editor/components/ColorPicker';
import { SliderPopover } from 'modules/common/components/ToolbarPopover';
import { convertDonutSizeChartDataToInput, convertDonutSizeInputToChartData } from './ProgressChartWidget.helpers';
import { ToolbarDivider } from 'modules/common/components/Toolbar/ToolbarDivider';
import { useEditor, useWidget } from 'widgets/sdk';

import { ReactComponent as CornerRadiusIcon } from 'assets/icons/corner_radius.svg';
import { ReactComponent as DonutHoleIcon } from 'assets/icons/donut_hole.svg';
import { ReactComponent as BarHeightIcon } from 'assets/icons/bar_height.svg';
import { MAX_BAR_HEIGHT, MIN_BAR_HEIGHT } from './ProgressChartWidget.config';
import { useTranslation } from 'react-i18next';

const DONUT_TYPES = [ProgressChartType.Donut, ProgressChartType.HalfDonut];

const BAR_HEIGHT_RANGE = MAX_BAR_HEIGHT - MIN_BAR_HEIGHT;

export const ProgressChartWidgetToolbar = () => {
  return (
    <Flex gap='2' align='center'>
      <DataColorOptions />
      <ToolbarDivider />
      <CornerRoundingOption />
      <DonutHoleOption />
      <BarHeightOption />
    </Flex>
  );
};

const DataColorOptions = () => {
  const { nonDataColor, dataColor, updateWidget } = useWidget<ProgressChartWidgetData>();
  const { t } = useTranslation('editor_widget_toolbar', { keyPrefix: 'progressChartToolbar', useSuspense: false });

  const onUpdateNonDataColor = (color: string) => updateWidget({ nonDataColor: color });
  const onUpdateDataColor = (color: string) => updateWidget({ dataColor: color });

  const nonDataLabel = t('nonDataColor.label');
  const dataLabel = t('dataColor.label');

  return (
    <>
      <ColorPicker color={nonDataColor} onChange={onUpdateNonDataColor} label={nonDataLabel} />
      <ColorPicker color={dataColor} onChange={onUpdateDataColor} label={dataLabel} />
    </>
  );
};

const CornerRoundingOption = () => {
  const { cornerRadius, updateWidget } = useWidget<ProgressChartWidgetData>();
  const { t } = useTranslation('editor_widget_toolbar', { keyPrefix: 'progressChartToolbar', useSuspense: false });

  const onUpdateCornerRadius = (radius: number) => updateWidget({ cornerRadius: radius / 100 });

  const label = t('cornerRadius.label');

  return (
    <SliderPopover
      title={label}
      label={label}
      icon={<CornerRadiusIcon />}
      value={Math.round(cornerRadius * 100)}
      min={0}
      max={100}
      suffix={'%'}
      onChange={onUpdateCornerRadius}
      activeIconStyle={{ '& path': { fill: 'upgrade.blue.700' } }}
    />
  );
};

const DonutHoleOption = () => {
  const { donutSize, type, updateWidget } = useWidget<ProgressChartWidgetData>();
  const { t } = useTranslation('editor_widget_toolbar', { keyPrefix: 'progressChartToolbar', useSuspense: false });

  const onUpdateDonutSize = (size: number) => updateWidget({ donutSize: convertDonutSizeInputToChartData(size, type) });

  const label = t('donutSize.label');

  /**
   * donutSize needs to be normalized due to issues rendering certain percentages
   * This will prevent setting donutSize to those percentages.
   */
  const donutSizeNormalized = useMemo(
    () => Math.round(convertDonutSizeChartDataToInput(donutSize, type)),
    [donutSize, type],
  );

  if (!DONUT_TYPES.includes(type)) return null;

  return (
    <SliderPopover
      title={label}
      label={label}
      icon={<DonutHoleIcon />}
      value={donutSizeNormalized}
      min={0}
      max={100}
      suffix={'%'}
      onChange={onUpdateDonutSize}
      activeIconStyle={{ '& path': { fill: 'upgrade.blue.700' } }}
    />
  );
};

const BarHeightOption = () => {
  const { type, barHeight, updateWidget, heightPx } = useWidget<ProgressChartWidgetData>();
  const { boundingBox } = useEditor();
  const { t } = useTranslation('editor_widget_toolbar', { keyPrefix: 'progressChartToolbar', useSuspense: false });

  const label = t('barHeight.label');

  const onUpdateBarHeight = (height: number) => {
    const scale = heightPx / barHeight;

    // barHeight is normalized to range between [5, 50] in the toolbar
    // Use min/max normalization to determine the heightPx from the new height
    updateWidget({
      barHeight: height,
      heightPx: (((height - MIN_BAR_HEIGHT) / BAR_HEIGHT_RANGE) * BAR_HEIGHT_RANGE + MIN_BAR_HEIGHT) * scale,
    });

    // Update BB after redux state update
    setTimeout(() => boundingBox.updateRect(), 0);
  };

  if (type !== ProgressChartType.Bar) return null;

  return (
    <SliderPopover
      title={label}
      label={label}
      icon={<BarHeightIcon />}
      value={barHeight}
      min={MIN_BAR_HEIGHT}
      max={MAX_BAR_HEIGHT}
      onChange={onUpdateBarHeight}
      activeIconStyle={{ '& path': { fill: 'upgrade.blue.700' } }}
    />
  );
};
