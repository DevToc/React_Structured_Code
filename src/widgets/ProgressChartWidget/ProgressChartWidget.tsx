import { useState, useCallback, useEffect } from 'react';
import { flushSync } from 'react-dom';
import { Box } from '@chakra-ui/react';

import { CustomOnResize } from 'modules/Editor/components/BoundingBox/BoundingBox.types';
import { CHART_TYPE_MAP } from './Charts';
import { DonutChartProps } from './Charts/Donut';
import { ProgressChartWidgetData } from './ProgressChartWidget.types';
import { ProgressChartWidgetToolbar } from './ProgressChartWidgetToolbar';
import { useWidget, ReadOnlyWidgetBase, WidgetBase, WidgetToolbar, useEditor } from 'widgets/sdk';
import { isSideHandle } from 'utils/boundingBox';

type Dimension = {
  widthPx: number;
  heightPx: number;
};

export const ProgressChartWidget = () => {
  const { widthPx, heightPx, donutSize } = useWidget<ProgressChartWidgetData>();
  const { boundingBox } = useEditor();
  const [dimension, setDimension] = useState<Dimension>({ widthPx, heightPx });

  useEffect(() => {
    boundingBox.updateRect();
  }, [donutSize, boundingBox]);

  /**
   * Listen for external changes to width + height
   * (i.e.) from undo/redo
   */
  useEffect(() => {
    setDimension({ widthPx, heightPx });
  }, [widthPx, heightPx]);

  /**
   * Custom onResize handler to update the chart dimensions
   * to allow for live rendering
   */
  const customOnResize = useCallback(
    ({ event, onResize, isGroup }: CustomOnResize) => {
      onResize(event);

      // Resize from side should keep same height
      const newHeight = isGroup && isSideHandle(event.direction) ? heightPx : event.target.clientHeight ?? event.height;
      event.target.style.height = `${newHeight}px`;

      // Use flushSync to avoid glitchy resizing due to React 18 batch updates
      flushSync(() =>
        setDimension({
          widthPx: event.target.clientWidth ?? event.width,
          heightPx: newHeight,
        }),
      );
    },
    [heightPx],
  );

  return (
    <WidgetBase onResize={customOnResize}>
      <WidgetToolbar>
        <ProgressChartWidgetToolbar />
      </WidgetToolbar>
      <ProgressChart dimension={dimension} />
    </WidgetBase>
  );
};

export const ReadOnlyProgressChartWidget = () => {
  const { widthPx, heightPx } = useWidget();

  return (
    <ReadOnlyWidgetBase disableSingleSelect>
      <ProgressChart dimension={{ widthPx, heightPx }} />
    </ReadOnlyWidgetBase>
  );
};

interface ProgressChartProps {
  dimension: Dimension;
}

const ProgressChart = ({ dimension }: ProgressChartProps) => {
  const { type, donutSize, value, cornerRadius, dataColor, nonDataColor } = useWidget<ProgressChartWidgetData>();
  const { widthPx, heightPx } = dimension;
  const { component: Chart, props: chartProps } = CHART_TYPE_MAP[type];

  return (
    <Box sx={{ '& svg': { position: 'absolute' } }}>
      <Chart
        type={'full'}
        donutSize={donutSize}
        value={value}
        cornerRadius={cornerRadius}
        dataColor={dataColor}
        nonDataColor={nonDataColor}
        widthPx={widthPx}
        heightPx={heightPx}
        {...(chartProps as Partial<DonutChartProps>)}
      />
    </Box>
  );
};
