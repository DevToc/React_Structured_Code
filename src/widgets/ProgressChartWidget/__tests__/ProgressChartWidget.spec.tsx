import { ProgressChartWidget } from '../ProgressChartWidget';
import { renderWidget } from 'utils/test-utils.test';
import { generateDefaultData } from '../ProgressChartWidget.helpers';
import { ProgressChartType } from '../ProgressChartWidget.types';
import { WidgetType } from 'types/widget.types';

const mockWidgetId = `${WidgetType.ProgressChart}-test-widget-id`;

describe('widgets/ProgressChartWidget', () => {
  it('should render donut correctly', () => {
    const widget = { widgetId: mockWidgetId, ...generateDefaultData(ProgressChartType.Donut) };
    const { asFragment } = renderWidget(<ProgressChartWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render half-donut correctly', () => {
    const widget = { widgetId: mockWidgetId, ...generateDefaultData(ProgressChartType.HalfDonut) };
    const { asFragment } = renderWidget(<ProgressChartWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render bar correctly', () => {
    const widget = { widgetId: mockWidgetId, ...generateDefaultData(ProgressChartType.Bar) };
    const { asFragment } = renderWidget(<ProgressChartWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });
});
