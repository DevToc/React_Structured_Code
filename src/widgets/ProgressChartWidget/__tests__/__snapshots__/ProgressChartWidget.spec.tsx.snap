// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/ProgressChartWidget should render bar correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="015-test-widget-id-description"
    aria-label=""
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-015-test-widget-id"
    id="015-test-widget-id"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 150px; height: 12px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-jmywsf"
    >
      <svg
        height="12"
        width="150"
      >
        <g>
          <rect
            fill="#D9E1FF"
            height="12"
            rx="6"
            ry="6"
            width="150"
            x="0"
            y="0"
          />
          <rect
            fill="#0069D1"
            height="12"
            rx="6"
            ry="6"
            width="100.5"
            x="0"
            y="0"
          />
        </g>
      </svg>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="015-test-widget-id-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/ProgressChartWidget should render donut correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="015-test-widget-id-description"
    aria-label=""
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-015-test-widget-id"
    id="015-test-widget-id"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 150px; height: 150px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-jmywsf"
    >
      <div
        style="position: relative;"
      >
        <svg
          height="150"
          role="img"
          width="150"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="transparent"
            height="150"
            width="150"
          />
          <g
            transform="translate(0,0)"
          >
            <g
              transform="translate(75,75)"
            >
              <path
                d="M4.592425496802574e-15,-75A75,75,0,1,1,-4.592425496802574e-15,75A75,75,0,1,1,4.592425496802574e-15,-75M-1.1986230546654718e-14,-65.25A65.25,65.25,0,1,0,1.1986230546654718e-14,65.25A65.25,65.25,0,1,0,-1.1986230546654718e-14,-65.25Z"
                fill="#D9E1FF"
              />
            </g>
            <g
              transform="translate(75,75)"
            >
              <path
                d="M6.217248937900877e-15,-69.95534289816611A4.875,4.875,0,0,1,5.213903743315515,-74.81854855418835A75,75,0,1,1,-63.052176604466474,40.61308933631091A4.875,4.875,0,0,1,-61.302334286422024,33.70124346423684L-61.302334286422024,33.70124346423684A4.875,4.875,0,0,1,-54.85539364588584,35.333387722590466A65.25,65.25,0,1,0,4.536096256684503,-65.09213724214386A4.875,4.875,0,0,1,6.217248937900877e-15,-69.95534289816611Z"
                fill="#0069D1"
                opacity="1"
                stroke="rgb(0, 74, 146)"
                stroke-width="0"
              />
              <path
                d="M-61.30233428642203,33.701243464236846A4.875,4.875,0,0,1,-68.07581117397636,31.475131977564722A75,75,0,0,1,-5.2139037433155275,-74.81854855418835A4.875,4.875,0,0,1,-1.3322676295501878e-14,-69.95534289816611L-1.3322676295501878e-14,-69.95534289816611A4.875,4.875,0,0,1,-4.5360962566845044,-65.09213724214386A65.25,65.25,0,0,0,-59.22595572135942,27.383364820481308A4.875,4.875,0,0,1,-61.30233428642203,33.701243464236846Z"
                fill="transparent"
                opacity="1"
                stroke="rgba(0, 0, 0, 0)"
                stroke-width="0"
              />
            </g>
          </g>
        </svg>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="015-test-widget-id-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/ProgressChartWidget should render half-donut correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="015-test-widget-id-description"
    aria-label=""
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-015-test-widget-id"
    id="015-test-widget-id"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 150px; height: 75px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-jmywsf"
    >
      <div
        style="position: relative;"
      >
        <svg
          height="75"
          role="img"
          width="150"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="transparent"
            height="75"
            width="150"
          />
          <g
            transform="translate(0,0)"
          >
            <g
              transform="translate(75,75)"
            >
              <path
                d="M-69.14658342969666,-1.0658141036401503e-14A5.625,5.625,0,0,1,-74.75306316723963,-6.081081081081093A75,75,0,0,1,74.75306316723963,-6.081081081081079A5.625,5.625,0,0,1,69.14658342969666,0L69.14658342969666,0A5.625,5.625,0,0,1,63.54010369215369,-5.1689189189189175A63.75,63.75,0,0,0,-63.540103692153686,-5.16891891891893A5.625,5.625,0,0,1,-69.14658342969666,-1.0658141036401503e-14Z"
                fill="#D9E1FF"
              />
            </g>
            <g
              transform="translate(75,75)"
            >
              <path
                d="M-69.14658342969666,-1.0658141036401503e-14A5.625,5.625,0,0,1,-74.75306316723963,-6.081081081081093A75,75,0,0,1,32.81816305022355,-67.43862523812997A5.625,5.625,0,0,1,35.19847472335396,-59.5173703816744L35.19847472335396,-59.5173703816744A5.625,5.625,0,0,1,27.89543859269001,-57.32283145241048A63.75,63.75,0,0,0,-63.540103692153686,-5.16891891891893A5.625,5.625,0,0,1,-69.14658342969666,-1.0658141036401503e-14Z"
                fill="#0069D1"
                opacity="1"
                stroke="rgb(0, 74, 146)"
                stroke-width="0"
              />
              <path
                d="M35.19847472335398,-59.51737038167439A5.625,5.625,0,0,1,43.28664716243369,-61.24758099251735A75,75,0,0,1,74.75306316723963,-6.081081081081063A5.625,5.625,0,0,1,69.14658342969666,1.5987211554602254e-14L69.14658342969666,1.5987211554602254e-14A5.625,5.625,0,0,1,63.54010369215369,-5.168918918918904A63.75,63.75,0,0,0,36.793650088068624,-52.060443843639746A5.625,5.625,0,0,1,35.19847472335398,-59.51737038167439Z"
                fill="transparent"
                opacity="1"
                stroke="rgba(0, 0, 0, 0)"
                stroke-width="0"
              />
            </g>
          </g>
        </svg>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="015-test-widget-id-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
